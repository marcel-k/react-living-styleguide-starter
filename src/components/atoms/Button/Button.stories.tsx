import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { text, select } from '@storybook/addon-knobs';

import Button from './Button';

storiesOf('atoms/Button', module)
  .add('default', () => (
    <Button
      onClick={action('button-click')}
    >
      {text('button-text', 'click me!')}
    </Button>
  ))
  .add('primary', () => (
    <Button
      primary
      onClick={action('button-click')}
    >
      {text('button-text', 'click me!')}
    </Button>
  ))
  .add('secondary', () => (
    <Button
      secondary
      onClick={action('button-click')}
    >
      {text('button-text', 'click me!')}
    </Button>
  ));