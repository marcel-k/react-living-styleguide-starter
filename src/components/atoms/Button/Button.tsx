import React from 'react';
import styled from 'styled-components/macro';

export interface ButtonProps {
  /**
   * Is this a primary button.
   * @default false
   */
  primary?: boolean;
  /**
   * Is this a secondary button.
   * @default false
   */
  secondary?: boolean;
}

//TODO: theme gebruiken
const color = (props: ButtonProps) =>
  props.primary ? '#fff' :
    props.secondary ? '#bbb' : 'inherit'; /* TODO: theme */

const backgroundColor = (props: ButtonProps) =>
  props.primary ? '#0000ff' :
    props.secondary ? 'transparent' : 'inherit'; /* TODO: theme */

const textDecoration = (props: ButtonProps) =>
  props.secondary ? 'underline' : 'none';

const border = (props: ButtonProps) =>
  props.secondary ? 0 : null;

// TODO: these use props from htmlElement, howto type these without breaking anything?
const opacity = ({ disabled }: any) => disabled ? 0.5 : 1;
const pointerEvents = ({ disabled }: any) => disabled ? 'none' : 'inherit';
const cursor = ({ disabled }: any) => disabled ? 'not-allowed' : 'pointer';

/**
 * A simple button component
 */
const Button = styled.button`
  height: 40px; /* TODO: theme */
  padding: 8px 16px; /* TODO: theme */
  font-size: 0.875rem; /* TODO: theme */
  text-align: center;

  :active {
    opacity: 0.5; /* TODO: theme */
  };

  opacity: ${opacity};
  cursor: ${cursor};
  pointer-events: ${pointerEvents};
  color: ${color};
  background-color: ${backgroundColor};
  text-decoration: ${textDecoration};
  font-variant: small-caps;
  border: ${border};
`;

export default Button;