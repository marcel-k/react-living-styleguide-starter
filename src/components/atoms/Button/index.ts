export * from './Button';
export { default as Button } from './Button';
export { default as ButtonRotatable } from './ButtonRotatable';