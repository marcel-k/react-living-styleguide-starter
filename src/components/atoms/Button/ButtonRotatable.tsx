import React from 'react';
import { useState, useEffect, useRef } from 'react';

import Button, { ButtonProps } from './Button';

interface Props extends ButtonProps {
  /**
   * Label of the button. Will be overwriten by any children of the component.
   */
  label?: string;
  /**
   * The direction the button will rotate.
   * @default left
   */
  direction?: 'left' | 'right';
  /**
   * fires when the button is clicked.
   */
  onClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>, currentRotation: number) => void;
  /**
   * If children is supplied, label will be overwritten by it.
   */
  children?: any;
}

/**
 * This button will rotate 90deg when you click on it.
 */
const ButtonRotatable = (props: Props) => {
  const {
    label,
    children,
    direction = 'left',
    onClick = () => undefined,
    ...buttonProps
  } = props;

  const buttonRef = useRef<HTMLButtonElement>(null);
  const [rotation, setRotation] = useState(0);

  useEffect(() => {
    if (buttonRef && buttonRef.current) {
      buttonRef.current.style.transform = `rotate(${rotation}deg)`;
    }
  }, [rotation]);

  const handleClick = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    let r = direction === 'left' ? rotation + 90 : rotation - 90;
    setRotation(r);
    onClick(event, r);
  };

  return (
    <Button
      onClick={handleClick}
      ref={buttonRef}
      {...buttonProps}
    >
      {children || label}
    </Button>
  );
};

export default ButtonRotatable;