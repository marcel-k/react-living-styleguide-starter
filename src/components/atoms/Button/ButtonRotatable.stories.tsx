import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { text, select } from '@storybook/addon-knobs';

import ButtonRotatable from './ButtonRotatable';

storiesOf('atoms/ButtonRotatable', module)
  .add('default', () => (
    <ButtonRotatable
      onClick={action('button-click')}
      label={text('button-text', 'click me!')}
    />
  ))
  .add('with children', () => (
    <ButtonRotatable
      label={'click me!'}
      onClick={action('button-click')}
    >
      <span>{text('button-text', 'click with children!')}</span>
    </ButtonRotatable>
  ))
  .add('reversed rotation', () => (
    <ButtonRotatable
      label={'click me!'}
      onClick={action('button-click')}
      direction={select('direction', ['left', 'right'], 'right')}
    >
    </ButtonRotatable>
  ));