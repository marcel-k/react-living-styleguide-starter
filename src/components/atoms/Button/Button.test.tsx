import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import 'jest-styled-components';
import 'react-testing-library/cleanup-after-each';

import Button from './Button';
import ButtonRotatable from './ButtonRotatable';

describe('<Button /> spec', () => {
  it('renders text as label', function () {
    const component = (<Button>click me!</Button>);
    const { container } = render(component);
    const { firstChild } = container;

    expect((firstChild as ChildNode).textContent).toEqual('click me!');
  });

  it('renders with primary colors', function () {
    const component = (<Button primary>click me!</Button>);
    const { container } = render(component);
    const { firstChild } = container;

    // TODO: get colors from theme? does this need to be tested like this?
    expect(firstChild).toHaveStyleRule('color', '#fff');
    expect(firstChild).toHaveStyleRule('background-color', '#0000ff');
  });

  it('renders with secondary colors', function () {
    const component = (<Button secondary>click me!</Button>);
    const { container } = render(component);
    const { firstChild } = container;

    // TODO: get colors from theme? does this need to be tested like this?
    expect(firstChild).toHaveStyleRule('color', '#bbb');
    expect(firstChild).toHaveStyleRule('text-decoration', 'underline');
  });
});

describe('<ButtonRotatable /> spec', () => {
  it('renders text as label', function () {
    const component = (<ButtonRotatable>click me!</ButtonRotatable>);
    const { container } = render(component);
    const { firstChild } = container;

    expect((firstChild as ChildNode).textContent).toEqual('click me!');
  });

  it('renders with children', function () {
    const component = (
      <ButtonRotatable label={`i won't be used`}>
        <span data-testid={'child'}>click me, I am a child!</span>
      </ButtonRotatable>
    );
    const { getByTestId } = render(component);

    expect(getByTestId('child')).toBeDefined();
  });

  it('rotates when clicked', function () {
    const component = (
      <ButtonRotatable direction={'left'}>
        I will rotate left
    </ButtonRotatable>
    );

    const { container, rerender } = render(component);

    const { firstElementChild } = container;
    expect(firstElementChild).toMatchSnapshot(); // 0deg

    fireEvent.click(firstElementChild as Element);
    expect(firstElementChild).toMatchSnapshot(); // 90deg

    rerender(
      <ButtonRotatable direction={'right'}>
        I will rotate right
      </ButtonRotatable>
    );

    fireEvent.click(firstElementChild as Element);
    expect(firstElementChild).toMatchSnapshot(); //0deg

    fireEvent.click(firstElementChild as Element);
    expect(firstElementChild).toMatchSnapshot(); //-90deg
  });
});
