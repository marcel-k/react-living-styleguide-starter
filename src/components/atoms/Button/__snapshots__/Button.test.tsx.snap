// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonRotatable /> spec rotates when clicked 1`] = `
.c0 {
  height: 40px;
  padding: 8px 16px;
  font-size: 0.875rem;
  text-align: center;
  opacity: 1;
  cursor: pointer;
  pointer-events: inherit;
  color: inherit;
  background-color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-variant: small-caps;
}

.c0:active {
  opacity: 0.5;
}

<button
  class="c0"
  style="transform: rotate(0deg);"
>
  I will rotate left
</button>
`;

exports[`<ButtonRotatable /> spec rotates when clicked 2`] = `
<button
  class="Button-qd808s-0 dClZAo"
  style="transform: rotate(90deg);"
>
  I will rotate left
</button>
`;

exports[`<ButtonRotatable /> spec rotates when clicked 3`] = `
<button
  class="Button-qd808s-0 dClZAo"
  style="transform: rotate(0deg);"
>
  I will rotate right
</button>
`;

exports[`<ButtonRotatable /> spec rotates when clicked 4`] = `
<button
  class="Button-qd808s-0 dClZAo"
  style="transform: rotate(-90deg);"
>
  I will rotate right
</button>
`;
